### PR TITLE
Working solidity merkle verifier

### PR DESCRIPTION
**Pull request by pvienhage**
- [ ] Tag the PR with `wip` while in development.
- [ ] Assign yourself as to the PR
- [ ] Assign relevant labels such as `bug`, `enhancement`.
- [ ] Request reviews if the PR is large, complex or you would like an extra pair of eyes to go over it.
- [ ] Add tests to cover changes as needed.
- [ ] Update documentation as needed.
- [ ] Add new entries to the `Changelog.md`.
- [ ] Update version numbers as needed.

Semver: https://github.com/rust-lang/rfcs/blob/master/text/1105-api-evolution.md

https://semver.org/

### CODE REVIEW

File: crypto/stark-verifier-ethereum/contracts/hashing.sol [hunk 1]
```diff
@@ -10,4 +12,8 @@ contract Hashing {
     function double_width_hash(bytes32 preimage_a, bytes32 preimage_b) internal pure returns(bytes32) {
         return hasher(abi.encodePacked(preimage_a, preimage_b));
     }
+
+    function double_width_masked_hash(bytes32 preimage_a, bytes32 preimage_b) internal pure returns(bytes32) {
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/hashing.sol]: The mask is part of the definition of our hash function, so it should not be exposed in the API of a hash-function abstraction.

If you need both masked and unmasked hash functions (I don't think we do?) then these are two different hash functions, and should have their own abstractions (which is awkward to do in Solidity, but better than a leaking abstraction)

File: crypto/stark-verifier-ethereum/contracts/merkle.sol [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        uint decommitment_index = 0;
+
+        while (true) {
+           (uint index, bytes32 current_hash) = remove_front(buffer);
+
+           uint parent = index/2;
+
+           if (parent > 0) {
+               // Checks if this is a left node
+               if (index%2 == 0) {
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/merkle.sol]: Spaces around `%`

File: crypto/stark-verifier-ethereum/contracts/merkle.sol [hunk 3]
```diff
@@ -0,0 +1,122 @@
+pragma solidity ^0.6.4;
+
+import './hashing.sol';
+import './ring_buffer.sol';
+
+contract MerkleVerifier is Hashing, RingBuffer {
+
+    function verify_merkle_proof(
+        bytes32 root,
+        bytes32[] memory data_points,
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/merkle.sol]: Can we call these `leaves` or do I misunderstand it?

File: crypto/stark-verifier-ethereum/contracts/merkle.sol [hunk 4]
```diff
@@ -0,0 +1,122 @@
+pragma solidity ^0.6.4;
+
+import './hashing.sol';
+import './ring_buffer.sol';
+
+contract MerkleVerifier is Hashing, RingBuffer {
+
+    function verify_merkle_proof(
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/merkle.sol]: Please add a line of documentation explaining how the arguments work, how indices are numbered and if they need to be supplied sorted.

File: crypto/stark-verifier-ethereum/contracts/merkle.sol [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+            is_empty: false
+        });
+        uint decommitment_index = 0;
+
+        while (true) {
+           (uint index, bytes32 current_hash) = remove_front(buffer);
+
+           uint parent = index/2;
+
+           if (parent > 0) {
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/merkle.sol]: Looks cleaner (less indented code) to turn this into an exit condition:

```
if (index == 1) {
    return(root == current_hash);
}
```

Doing it directly on `index` avoids creating an invalid `parent` index.

File: crypto/stark-verifier-ethereum/contracts/merkle.sol (hunk 2 above)
Comment (Recmo) [reply crypto/stark-verifier-ethereum/contracts/merkle.sol]: Can we assign this to an `is_left` variable and use it in places where we now hardcode `true`?

File: crypto/stark-verifier-ethereum/contracts/merkle.sol [hunk 6]
```diff
@@ -0,0 +1,122 @@
+pragma solidity ^0.6.4;
+
+import './hashing.sol';
+import './ring_buffer.sol';
+
+contract MerkleVerifier is Hashing, RingBuffer {
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/merkle.sol]: Solidity has something like `use RingBuffer for IndexRingBuffer` to allow member function syntax. This would be a good place to use it.

File: crypto/stark-verifier-ethereum/contracts/merkle.sol [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    // It returns true if decommitnet data exists and false if it doesn't.
+    function read_decommitment_and_push(
+        bool is_left,
+        IndexRingBuffer memory buffer,
+        bytes32[] memory decommitment,
+        uint decommitment_index,
+        bytes32 current_hash,
+        uint index) internal pure returns(bool) {
+            // We do not want a revert from reading too far into the decommitment.
+            if (decommitment_index < decommitment.length) {
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/merkle.sol]: Turn this if into an early exit.

File: crypto/stark-verifier-ethereum/contracts/ring_buffer.sol [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+            uint next = (buffer.back + 1)%buffer.data.length;
+            buffer.data[next] = data;
+            buffer.indexes[next] = index;
+            buffer.back = next;
+        }
+    }
+
+    // Removes an element from the front of the buffer, and moves the front forward.
+    function remove_front(IndexRingBuffer memory buffer) internal pure returns(uint index, bytes32 data) {
+        // If the buffer is empty return 0, 0
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/ring_buffer.sol]: Why `(0,0)` and not revert?

File: crypto/stark-verifier-ethereum/contracts/ring_buffer.sol [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        // Otherwise we push and move the back index mod the length.
+        if (buffer.is_empty) {
+            buffer.front = 0;
+            buffer.back = 0;
+            buffer.is_empty = false;
+            buffer.indexes[0] = index;
+            buffer.data[0] = data;
+        } else {
+            // We could add a check here that the push doesn't put the buffer into
+            // the state where front = back, but we don't want reverts here.
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/ring_buffer.sol]: Why don't we want reverts here?

File: crypto/stark-verifier-ethereum/contracts/merkle.sol [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+    // This function reads from decomitment and pushes the new node onto the buffer,
+    // It returns true if decommitnet data exists and false if it doesn't.
+    function read_decommitment_and_push(
+        bool is_left,
+        IndexRingBuffer memory buffer,
+        bytes32[] memory decommitment,
+        uint decommitment_index,
+        bytes32 current_hash,
+        uint index) internal pure returns(bool) {
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/merkle.sol]: This should go on two lines, like in the other functions.

File: crypto/stark-verifier-ethereum/test/merkle_test.ts [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+  let merkle_contract: any;
+  const init_hex = "0x0123456789abcded";
+
+  const provider = waffle.provider;
+  let [wallet] = provider.getWallets();
+
+  before(async function() {
+    merkle_contract = await deployContract(wallet, MerkleVerifierTestArtifact) as MerkleVerifierTest;
+  });
+
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/test/merkle_test.ts]: Can we add a test where all leaves are provided and the decommitment is empty?

File: crypto/stark-verifier-ethereum/test/merkle_test.ts [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+describe("Public coin testing", () => {
+  let merkle_contract: any;
+  const init_hex = "0x0123456789abcded";
+
+  const provider = waffle.provider;
+  let [wallet] = provider.getWallets();
+
+  before(async function() {
+    merkle_contract = await deployContract(wallet, MerkleVerifierTestArtifact) as MerkleVerifierTest;
+  });
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/test/merkle_test.ts]: Can we add a test where clamed_data is empty? What should the correct behaviour be?

File: crypto/stark-verifier-ethereum/contracts/merkle.sol [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+                    decommitment_iter,
+                    current_hash,
+                    index
+                );
+            }
+        }
+    }
+
+    // This function reads from decomitment and pushes the new node onto the buffer,
+    // It returns true if decommitnet data exists and false if it doesn't.
```
Comment (Recmo) [inline crypto/stark-verifier-ethereum/contracts/merkle.sol]: Spelling
